crates/corro-agent/src/api/public/update.rs

```diff
@@ -0,0 +1,284 @@
+        match upsert_update(handle.clone(), maybe_created, updates, &mut bcast_write).await {
+            Ok(id) => id,
+            Err(e) => return hyper::Response::<hyper::Body>::from(e),
+        };
+
+    tokio::spawn(forward_update_bytes_to_body_sender(
+        handle, sub_rx, tx, tripwire,
+    ));
+
+    hyper::Response::builder()
+        .status(StatusCode::OK)
+        .header("corro-query-id", update_id.to_string())
+        .body(body)
+        .expect("could not generate ok http response for update request")
+}
+
+pub async fn upsert_update(
+    handle: UpdateHandle,
+    maybe_created: Option<UpdateCreated>,
+    updates: &UpdatesManager,
+    bcast_write: &mut UpdateBroadcastCache,
+) -> Result<(Uuid, broadcast::Receiver<Bytes>), MatcherUpsertError> {
+    let sub_rx = if let Some(created) = maybe_created {
+        let (sub_tx, sub_rx) = broadcast::channel(10240);
+        bcast_write.insert(handle.id(), sub_tx.clone());
+        tokio::spawn(process_update_channel(
+            updates.clone(),
+            handle.id(),
+            sub_tx,
+            created.evt_rx,
+        ));
+
+        sub_rx
+    } else {
+        let id = handle.id();
+        let sub_tx = bcast_write
+            .get(&id)
+            .cloned()
+            .ok_or(MatcherUpsertError::MissingBroadcaster)?;
+        debug!("found update handle");
+
+        sub_tx.subscribe()
+    };
+
+    Ok((handle.id(), sub_rx))
+}
+
+pub async fn process_update_channel(
+    updates: UpdatesManager,
+    id: Uuid,
+    tx: broadcast::Sender<Bytes>,
+    mut evt_rx: mpsc::Receiver<NotifyEvent>,
+) {
+    let mut buf = BytesMut::new();
+
+    let mut deadline = if tx.receiver_count() == 0 {
+        Some(Box::pin(tokio::time::sleep(MAX_UNSUB_TIME)))
+    } else {
+        None
+    };
+
+    // even if there are no more subscribers
+    // useful for queries that don't change often so we can cleanup...
+    let mut subs_check = tokio::time::interval(RECEIVERS_CHECK_INTERVAL);
+
+    loop {
+        let deadline_check = async {
+            if let Some(sleep) = deadline.as_mut() {
+                sleep.await
+            } else {
+                futures::future::pending().await
+            }
+        };
+
+        let notify_evt = tokio::select! {
+            biased;
+            Some(query_evt) = evt_rx.recv() => query_evt,
+            _ = deadline_check => {
+                if tx.receiver_count() == 0 {
+                    info!(sub_id = %id, "All listeners for subscription are gone and didn't come back within {MAX_UNSUB_TIME:?}");
+                    break;
+                }
+
+                // reset the deadline if there are receivers!
+                deadline = None;
+                continue;
+            },
+            _ = subs_check.tick() => {
+                if tx.receiver_count() == 0 {
+                    if deadline.is_none() {
+                        deadline = Some(Box::pin(tokio::time::sleep(MAX_UNSUB_TIME)));
```

Comment:
I don't think we need the whole deadline check because it's stateless. As soon as update listeners are gone, we should just remove it.

It should simplify the code significantly here.